querqy-core/src/main/java/querqy/rewrite/lookup/StateExchangingSequenceExtractor.java

```diff
@@ -0,0 +1,126 @@
+package querqy.rewrite.lookup;
+
+import lombok.RequiredArgsConstructor;
+import querqy.model.AbstractNodeVisitor;
+import querqy.model.BooleanClause;
+import querqy.model.BooleanQuery;
+import querqy.model.Term;
+import querqy.rewrite.lookup.model.Sequence;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+
+@RequiredArgsConstructor(staticName = "of")
+public class StateExchangingSequenceExtractor<T> extends AbstractNodeVisitor<Void> {
+
+    protected static final Term BOUNDARY_TERM = new Term(null, "\u0002");
+
+    private final BooleanQuery booleanQuery;
+    private final StateExchangingCollector<T, ?> stateExchangingCollector;
+    private final boolean hasBoundaries;
+
+    private List<Sequence<T>> previousSequences = List.of();
+    private List<Sequence<T>> sequences = new ArrayList<>();
+
+    public void extractSequences() {
+        potentiallyEvaluateBoundaryTerm();
+        visitBooleanQuery();
+        potentiallyEvaluateBoundaryTerm();
+    }
+
+    private void potentiallyEvaluateBoundaryTerm() {
+        if (hasBoundaries) {
+            visit(BOUNDARY_TERM);
+            refreshSequenceLists();
+        }
+    }
+
+    private void visitBooleanQuery() {
+        for (final BooleanClause clause : booleanQuery.getClauses()) {
+            clause.accept(this);
+            refreshSequenceLists();
+        }
+    }
+
+    private void refreshSequenceLists() {
+        if (!sequences.isEmpty()) {
+            previousSequences = sequences;
+            sequences = new ArrayList<>();
+
+        } else if (!previousSequences.isEmpty()) {
+            previousSequences = new ArrayList<>();
+        }
+    }
+
+    @Override
+    public Void visit(final BooleanQuery booleanQuery) {
+        // TODO: return all full sequences from bq to enable cross-hierarchy lookups
```

Comment:
Is this still to be done before the merge?

Reply:
Depends on the remaining time. However, cross-hierarchical lookups are currently not done in Querqy so I think it is optional. Would be beneficial e.g. for subsequent decompounding and synonym lookups.